### PR TITLE
Provides a main framework for custom scripting

### PR DESCRIPTION
**Pull request by andyHa**
### Description

Permits to handle custom events on a "per tenant" level. For now, this only provides a MongoDB based repository.

### Checklist

- [x] Code change has been tested and works locally
- [x] Code was formatted via IntelliJ and follows SonarLint & [best practices](https://scireum.myjetbrains.com/youtrack/articles/MISC-A-16/CodeStyle-JavaDoc)
- [x] Patch Tasks: Is local execution of Patch Tasks necessary? If so, please also mark the PR with the tag.

### CODE REVIEW

File: src/main/java/sirius/biz/scripting/CustomEvent.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+package sirius.biz.scripting;
+
+import sirius.kernel.health.HandledException;
+
+import java.util.Optional;
+
+/**
+ * Provides a base class for all custom events handled by a {@link CustomEventDispatcher}.
+ */
+public abstract class CustomEvent {
```
Comment (sabieber) [inline src/main/java/sirius/biz/scripting/CustomEvent.java:18]: **Optional:** In my opinion the class name itself (and also `CustomEventDispatcher`) doesn't reflect that it is related to the scripting capabilities and could be anything. I would prefer something like `ScriptingEvent`.

File: src/main/java/sirius/biz/scripting/CustomEvent.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    protected boolean success;
+    protected boolean failed;
+    protected HandledException error;
```
Comment (sabieber) [inline src/main/java/sirius/biz/scripting/CustomEvent.java:22]: **Optional:** Would like the JavaDoc of the methods on the fields themselves, as this helps when implementing or debugging subclasses of this abstract class.

File: src/main/java/sirius/biz/scripting/ScriptBasedCustomEventDispatcher.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                                                              ignored -> handleExceptionInProcess(handledException));
+        }
+    }
+
+    private void handleExceptionInProcess(HandledException handledException) {
+        TaskContext.get()
+                   .log(NLS.fmtr("CustomEventHandler.message")
+                           .set("system", TaskContext.get().getSystemString())
+                           .set("message", handledException.getMessage())
+                           .format());
```
Comment (sabieber) [inline src/main/java/sirius/biz/scripting/ScriptBasedCustomEventDispatcher.java:109]: **Optional:** Arent there logging capabilities where a NLS key and parameters can be supplied directly so the message is not stored translated but only translated in the correct display language when viewed?

File: src/main/java/sirius/biz/scripting/mongo/MongoCustomEventDispatcherRepository.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Stores and manages {@link CustomEventDispatcher custom event dispatchers} in a MongoDB.
+ */
+@Register(framework = MongoCustomEventDispatcherRepository.FRAMEWORK_SCRIPTING_MONGO)
+public class MongoCustomEventDispatcherRepository implements CustomEventDispatcherRepository {
+
+    protected static final String FRAMEWORK_SCRIPTING_MONGO = "biz.scripting-mongo";
```
Comment (sabieber) [inline src/main/java/sirius/biz/scripting/mongo/MongoCustomEventDispatcherRepository.java:41]: **Question:** Is this a new framework? JavaDoc may be nice in that case.

File: src/main/java/sirius/biz/scripting/mongo/MongoCustomEventDispatcherRepository.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            NoodleCompiler compiler = new NoodleCompiler(compilationContext);
+            Callable compiledScript = compiler.compileScript();
+
+            ScriptBasedCustomEventDispatcher dispatcher = new ScriptBasedCustomEventDispatcher();
+            SimpleEnvironment environment = new SimpleEnvironment();
+            environment.writeVariable(0, dispatcher);
+            compiledScript.call(environment);
+
+            return Optional.of(dispatcher);
+        } catch (ScriptingException | HandledException e) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/scripting/mongo/MongoCustomEventDispatcherRepository.java:101]: **Optional:** Rename `e` to `exception` :)

File: src/main/java/sirius/biz/scripting/mongo/MongoCustomScriptController.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Part
+    private Mango mango;
+
+    /**
+     * Lists all scripts available for the current tenant.
+     *
+     * @param webContext the request to handle
+     */
+    @Routed("/scripting/scripts")
```
Comment (sabieber) [inline src/main/java/sirius/biz/scripting/mongo/MongoCustomScriptController.java:47]: **Question:** Are these included in the menu-bar somewhere?

File: src/main/java/sirius/biz/scripting/mongo/MongoCustomScript.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import sirius.db.mixing.Mapping;
+import sirius.db.mixing.annotations.NullAllowed;
+import sirius.db.mixing.annotations.Unique;
+import sirius.kernel.commons.Strings;
+import sirius.kernel.nls.NLS;
+
+/**
+ * Stores a custom scripting profile for a tenant.
+ */
+public class MongoCustomScript extends MongoTenantAware {
```
Comment (idlira) [inline src/main/java/sirius/biz/scripting/mongo/MongoCustomScript.java:24]: Add a composite index tenant + code

**Review by idlira — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/biz/scripting/CustomEvent.java (hunk 2 above)
Comment (andyHa) [reply src/main/java/sirius/biz/scripting/CustomEvent.java]: done

File: src/main/java/sirius/biz/scripting/ScriptBasedCustomEventDispatcher.java (hunk 3 above)
Comment (andyHa) [reply src/main/java/sirius/biz/scripting/ScriptBasedCustomEventDispatcher.java]: This API is not exposed via TaskContext and only available in Processes. While a Process will most probably be around, we want to decouple the two frameworks and rely on the more generic API of TaskContext here...

File: src/main/java/sirius/biz/scripting/mongo/MongoCustomEventDispatcherRepository.java (hunk 4 above)
Comment (andyHa) [reply src/main/java/sirius/biz/scripting/mongo/MongoCustomEventDispatcherRepository.java]: done

File: src/main/java/sirius/biz/scripting/mongo/MongoCustomEventDispatcherRepository.java (hunk 5 above)
Comment (andyHa) [reply src/main/java/sirius/biz/scripting/mongo/MongoCustomEventDispatcherRepository.java]: done

File: src/main/java/sirius/biz/scripting/mongo/MongoCustomScriptController.java (hunk 6 above)
Comment (andyHa) [reply src/main/java/sirius/biz/scripting/mongo/MongoCustomScriptController.java:47]: done

File: src/main/java/sirius/biz/scripting/mongo/MongoCustomScript.java (hunk 7 above)
Comment (andyHa) [reply src/main/java/sirius/biz/scripting/mongo/MongoCustomScript.java:27]: done

**Review by sabieber — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Pull request merged**